src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,129 @@
+    }
+
     @Override
     public void add(T value) {
-
+        if (storage.length == size) {
+            increaseSize();
+        }
+        storage[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (!adding(value, index)) {
+            throw new ArrayListIndexOutOfBoundsException(OF_BOUND_MESSAGE);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list != null) {
+            for (int i = 0; i < list.size(); i++) {
+                add(list.get(i));
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        T result = getting(index);
+        if (result != null && result.getClass().getGenericSuperclass() == null) {
+            throw new ArrayListIndexOutOfBoundsException(OF_BOUND_MESSAGE);
+        }
+        return result;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!setting(value, index)) {
+            throw new ArrayListIndexOutOfBoundsException(OF_BOUND_MESSAGE);
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T result = removing(index);
+        if (result == null) {
+            throw new ArrayListIndexOutOfBoundsException(OF_BOUND_MESSAGE);
+        }
+        return result;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(storage[i], element)) {
+                return removing(i);
+            }
+        }
+        throw new NoSuchElementException(OF_ABSENT_MESSAGE);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean adding(T value, int index) {
+        if (index >= 0 && index <= size) {
+            List<T> tempList = null;
+            if (index < size) {
+                tempList = new ArrayList<>();
+                for (int i = index; i < size; i++) {
+                    tempList.add(get(i));
+                }
+            }
+            size = index;
+            add(value);
+            addAll(tempList);
+        }
+        return index >= 0 && index < size;
+    }
+
+    private T getting(int index) {
+        return index < 0 || index >= size ? (T) new Object() : storage[index];
+    }
+
+    private T removing(int index) {
+        T result = null;
+        if (index >= 0 && index < size) {
+            result = storage[index];
+            System.arraycopy(storage, index + 1, storage, index, size - index - 1);
+            storage[--size] = null;
+        }
+        return result;
+    }
+
+    private boolean setting(T value, int index) {
+        if (index >= 0 && index < size) {
+            storage[index] = value;
+        }
+        return index >= 0 && index < size;
+    }
```

Comment:
in general, most of your private methods should be removed and the main business logic from them should be moved to its corresponding public methods, except for `increaseSize()` `checkIndex()`

Reply:
refactored